src/common/internet.rs

```diff
@@ -0,0 +1,310 @@
+// LNP/BP Rust Library
+// Written in 2020 by
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+///! Universal addresses that support IPv4, IPv6 and Tor
+
+use std::fmt;
+use std::convert::TryFrom;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+#[cfg(feature="use-tor")]
+use torut::onion::{TorPublicKeyV3, OnionAddressV3, TORV3_PUBLIC_KEY_LENGTH};
+
+
+/// A universal address covering IPv4, IPv6 and Tor in a single byte sequence
+/// of 32 bytes.
+///
+/// Holds either:
+/// * IPv4-to-IPv6 address
+/// * IPv6 address
+/// * Tor address (only 3rd version is supported)
+///
+/// NB: we are using `TorPublicKeyV3` instead of `OnionAddressV3`, since
+/// `OnionAddressV3` keeps cehcksum and other information wich can be
+/// reconstructed from `TorPublicKeyV3`. The 2-byte checksum in `OnionAddressV3`
+/// is designed for human-readable part that checks that the address was typed
+/// in correctly. In computer-stored digital data it may be deterministically
+/// regenerated and does not add any additional security.
+///
+/// Tor addresses are distinguished by the fact that last 16 bits
+/// must be set to 0
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum InetAddr {
+    IPv4(Ipv4Addr),
+    IPv6(Ipv6Addr),
+    #[cfg(feature="use-tor")]
+    Tor(TorPublicKeyV3),
+}
+
+impl InetAddr {
+    pub fn get_ip6(&self) -> Option<Ipv6Addr> {
+        match self {
+            InetAddr::IPv4(ipv4_addr) => Some(ipv4_addr.to_ipv6_mapped()),
+            InetAddr::IPv6(ipv6_addr) => Some(*ipv6_addr),
+            #[cfg(feature="use-tor")]
+            _ => None,
+        }
+    }
+
+    pub fn from_32bit_encoding(data: [u8; TORV3_PUBLIC_KEY_LENGTH]) -> Option<Self> {
```

Comment:
"32byte"

Reply:
Decided to name it in other fashion `from_uniform_encoding`